actions/generate-schedule/scheduler/scheduler.go

```diff
@@ -0,0 +1,98 @@
+package scheduler
+
+import (
+	"fmt"
+	"github.com/spinnaker/rotation-scheduler/actions/generate-schedule/users"
+	"github.com/spinnaker/rotation-scheduler/proto/schedule"
+	"time"
+)
+
+const (
+	DateFormat = "Mon 02 Jan 2006"
+)
+
+// Scheduler creates or extends a new rotation schedule.
+type Scheduler struct {
+	userSource        users.Source
+	shiftDurationDays int
+}
+
+// NewScheduler creates a new Scheduler. All args are required.
+func NewScheduler(userSource users.Source, shiftDurationDays int) (*Scheduler, error) {
+	if userSource == nil {
+		return nil, fmt.Errorf("no user source specificed")
+	}
+
+	if shiftDurationDays <= 0 {
+		return nil, fmt.Errorf("shift duration invalid. Must be greater than 0, was: %v", shiftDurationDays)
+	}
+
+	return &Scheduler{
+		userSource:        userSource,
+		shiftDurationDays: shiftDurationDays,
+	}, nil
+}
+
+// Schedule creates a new Schedule from start (inclusive) to stop (exclusive). Will return a 1-entry schedule if
+// stop is before start.
+func (s *Scheduler) Schedule(start, stop time.Time) (*schedule.Schedule, error) {
+	return s.ExtendSchedule(&schedule.Schedule{
+		Shifts: []*schedule.Shift{{
+			StartDate: start.Format(DateFormat),
+			User:      s.userSource.NextUser(),
+		}},
+	}, stop)
+}
+
+// ExtendSchedule takes a previously generated schedule and extends it. The user rotation continues as normal from the
+// last shift in the schedule.
+func (s *Scheduler) ExtendSchedule(previousSchedule *schedule.Schedule, stop time.Time) (*schedule.Schedule, error) {
+	if err := validatePreviousShifts(previousSchedule.Shifts); err != nil {
+		return nil, fmt.Errorf("error validating previous shifts: %v", err)
+	}
+
+	newSchedule := &schedule.Schedule{
+		Shifts: previousSchedule.Shifts,
+	}
+
+	mostRecentShift := previousSchedule.Shifts[len(previousSchedule.Shifts)-1]
+	if err := s.userSource.StartAfter(mostRecentShift.User); err != nil {
+		return nil, fmt.Errorf("error finding previous shift owner (%v) in user source: %v", mostRecentShift.User, err)
+	}
+
+	for start := s.startTime(mostRecentShift); start.Before(stop); start = start.AddDate(0, 0, s.shiftDurationDays) {
+		newSchedule.Shifts = append(newSchedule.Shifts, &schedule.Shift{
+			User:      s.userSource.NextUser(),
+			StartDate: start.Format(DateFormat),
+		})
+	}
+
+	return newSchedule, nil
+}
+
+func validatePreviousShifts(previousShifts []*schedule.Shift) error {
+	if previousShifts == nil || len(previousShifts) == 0 {
+		fmt.Printf("No previous shifts found. Starting from scratch.")
+		return nil
+	}
+
+	for i, shift := range previousShifts {
```

Comment:
Fair question - nothing is _requiring_ it, but I'd say we would need to prior to any updates to a shared calendar. So it'd be either do it now or during the calendar sync phase, WDYT?

---

> but it will be annoying if at some point in the future we have 600 shifts and one in the past fails validation for some reason.

I intend to add some logic to prune entries from the past.

Reply:
Yeah, if this is pushing to a shared calendar, I'd agree we should validate each entry before updating the shared calendar.  If we'll always be extending and syncing together, then I suppose it doesn't really matter where the validation is...but maybe it's a bit cleaner if it's closer to where we need the data to be valid (ie in the sync logic)?

Reply:
I'll move the validation logic to the calendar sync when it gets written, but leave it here until then.